File: rust/present80/math.rs
pub(crate) fn bit_ones(size: u128) -> u128 {
    let mut num = 0x00;

    for i in 0..size {
        num |= 1 << i;
    }

    num
}

pub(crate) fn rotate_right(num: u128, bits: u128, width: u128) -> u128 {
    const MAX_WIDTH: u128 = 128;

    let width = if width > MAX_WIDTH { MAX_WIDTH } else { width };

    let bits = if bits > MAX_WIDTH {
        bits % MAX_WIDTH
    } else {
        bits
    };

    ((num & bit_ones(bits)) << (width - bits)) | (num >> bits)
}

File: c/present80/math.c
#include "math.h"

u128 bit_ones(u128 size)
{
	u128 num = 0x00;

	for (size_t i = 0; i < size; i++)
		num |= 1 << i;

	return num;
}

u128 rotate_right(u128 num, u8 bits, u8 width)
{
    const u8 MAX_WIDTH = 128;

    if(width > MAX_WIDTH)
        width = MAX_WIDTH;

    if(bits > width)
        bits = bits % width;

    return ((num & bit_ones(bits)) << (width - bits)) | (num >> bits);
}
